From 0db206172b54b30e1bc35d5125f82b22d96f20a8 Mon Sep 17 00:00:00 2001
From: Miroslav Ondra <ondra@faster.cz>
Date: Mon, 4 Apr 2022 16:29:07 +0200
Subject: [PATCH 03/12] Add rxtx leds for imx serial.

---
 drivers/tty/serial/Kconfig | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/drivers/tty/serial/Kconfig b/drivers/tty/serial/Kconfig
index 131a6a587..d7692c5cd 100644
--- a/drivers/tty/serial/Kconfig
+++ b/drivers/tty/serial/Kconfig
@@ -1561,6 +1561,15 @@ config SERIAL_LITEUART_CONSOLE
 	  and warnings and which allows logins in single user mode).
 	  Otherwise, say 'N'.
 
+config RXTX_LED_TRIG
+	bool "RXTX Serial LED Triggers"
+	depends on LEDS_CLASS && LEDS_TRIGGERS && (SERIAL_IMX || SERIAL_8250_DW)
+	help
+	  This option adds LED triggers for RX/TX serial activity.
+
+	  Say Y here if you are working on a system with led-class supported
+	  LEDs and you want to use them as activity indicators for serial port.
+
 endmenu
 
 config SERIAL_MCTRL_GPIO
-- 
2.20.1

