From d39adc05b5918ca80547f2a2552a862df27f0212 Mon Sep 17 00:00:00 2001
From: Miroslav Ondra <ondra@faster.cz>
Date: Mon, 19 Feb 2024 10:47:09 +0100
Subject: [PATCH 4/7] Add /sys/devices/soc0/soc_uid to be compatible with

---
 drivers/soc/imx/soc-imx8m.c | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/drivers/soc/imx/soc-imx8m.c b/drivers/soc/imx/soc-imx8m.c
index 5b823d58a..f360e1e48 100644
--- a/drivers/soc/imx/soc-imx8m.c
+++ b/drivers/soc/imx/soc-imx8m.c
@@ -48,6 +48,14 @@ struct imx8_soc_data {
 static u64 soc_uid;
 static u64 soc_uid_h;
 
+static ssize_t soc_uid_show(struct device *dev,
+			    struct device_attribute *attr, char *buf)
+{
+	return sprintf(buf, "%016llX\n", soc_uid);
+}
+
+static DEVICE_ATTR_RO(soc_uid);
+
 #ifdef CONFIG_HAVE_ARM_SMCCC
 static u32 imx8mq_soc_revision_from_atf(void)
 {
@@ -265,6 +273,11 @@ static int __init imx8_soc_init(void)
 		goto free_serial_number;
 	}
 
+	ret = device_create_file(soc_device_to_device(soc_dev),
+				 &dev_attr_soc_uid);
+	if (ret)
+		goto free_serial_number;
+
 	pr_info("SoC: %s revision %s\n", soc_dev_attr->soc_id,
 		soc_dev_attr->revision);
 
-- 
2.30.2

