From 4eadb52168b473716dff0741e1d5883c71612f1f Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Martin=20T=C5=99=C3=ADska?= <triska@unipi.technology>
Date: Tue, 21 Sep 2021 11:53:29 +0000
Subject: [PATCH 05/10] Add support for Quectel EG912Y

---
 drivers/usb/serial/option.c | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/drivers/usb/serial/option.c b/drivers/usb/serial/option.c
index c6969ca72..47af79e75 100644
--- a/drivers/usb/serial/option.c
+++ b/drivers/usb/serial/option.c
@@ -251,6 +251,8 @@ static void option_instat_callback(struct urb *urb);
 #define QUECTEL_PRODUCT_EM12			0x0512
 #define QUECTEL_PRODUCT_RM500Q			0x0800
 #define QUECTEL_PRODUCT_EC200T			0x6026
+#define QUECTEL_PRODUCT_EG912Y			0x6001
+#define QUECTEL_PRODUCT_EC100Y			0x6000
 
 #define CMOTECH_VENDOR_ID			0x16d8
 #define CMOTECH_PRODUCT_6001			0x6001
@@ -1126,6 +1128,9 @@ static const struct usb_device_id option_ids[] = {
 	{ USB_DEVICE_AND_INTERFACE_INFO(QUECTEL_VENDOR_ID, QUECTEL_PRODUCT_RM500Q, 0xff, 0xff, 0x10),
 	  .driver_info = ZLP },
 	{ USB_DEVICE_AND_INTERFACE_INFO(QUECTEL_VENDOR_ID, QUECTEL_PRODUCT_EC200T, 0xff, 0, 0) },
+	{USB_DEVICE(QUECTEL_VENDOR_ID, QUECTEL_PRODUCT_EG912Y)},
+	{USB_DEVICE(QUECTEL_VENDOR_ID, QUECTEL_PRODUCT_EC100Y)},
+
 
 	{ USB_DEVICE(CMOTECH_VENDOR_ID, CMOTECH_PRODUCT_6001) },
 	{ USB_DEVICE(CMOTECH_VENDOR_ID, CMOTECH_PRODUCT_CMU_300) },
@@ -2148,6 +2153,13 @@ static int option_probe(struct usb_serial *serial,
 	if (device_flags & NUMEP2 && iface_desc->bNumEndpoints != 2)
 		return -ENODEV;
 
+	if(serial->dev->descriptor.idVendor == cpu_to_le16(QUECTEL_VENDOR_ID)){
+		__u16 idProduct = le16_to_cpu(serial->dev->descriptor.idProduct);
+
+		if (serial->interface->cur_altsetting->desc.bInterfaceClass != 0xFF)
+			return -ENODEV;
+	}
+
 	/* Store the device flags so we can use them during attach. */
 	usb_set_serial_data(serial, (void *)device_flags);
 
-- 
2.20.1

