From 896b95a1c43950dd00d81bf6d1e8f37945bde228 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Martin=20T=C5=99=C3=ADska?= <triska@unipi.technology>
Date: Mon, 15 Nov 2021 13:32:24 +0000
Subject: [PATCH 07/10] Extend Quectel EG912Y support by pid 0x6002

---
 drivers/usb/serial/option.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/drivers/usb/serial/option.c b/drivers/usb/serial/option.c
index 7d8043907..a31429d8b 100644
--- a/drivers/usb/serial/option.c
+++ b/drivers/usb/serial/option.c
@@ -252,6 +252,7 @@ static void option_instat_callback(struct urb *urb);
 #define QUECTEL_PRODUCT_RM500Q			0x0800
 #define QUECTEL_PRODUCT_EC200T			0x6026
 #define QUECTEL_PRODUCT_EG912Y			0x6001
+#define QUECTEL_PRODUCT_EG912Y_EU		0x6002
 #define QUECTEL_PRODUCT_EC100Y			0x6000
 
 #define CMOTECH_VENDOR_ID			0x16d8
@@ -1129,6 +1130,7 @@ static const struct usb_device_id option_ids[] = {
 	  .driver_info = ZLP },
 	{ USB_DEVICE_AND_INTERFACE_INFO(QUECTEL_VENDOR_ID, QUECTEL_PRODUCT_EC200T, 0xff, 0, 0) },
 	{USB_DEVICE(QUECTEL_VENDOR_ID, QUECTEL_PRODUCT_EG912Y)},
+	{USB_DEVICE(QUECTEL_VENDOR_ID, QUECTEL_PRODUCT_EG912Y_EU)},
 	{USB_DEVICE(QUECTEL_VENDOR_ID, QUECTEL_PRODUCT_EC100Y)},
 
 
-- 
2.20.1

